### PR TITLE
Don't crash if a Lua error occurs inside get_staticdata

### LINKED ISSUE #13620: get_staticdata errors are not handeled (crash to desktop)
##### Minetest version

```
5.7.0
```

##### Steps to reproduce

1. Load a mod with the following code:
```lua
minetest.register_entity(":crash_test", {
    get_staticdata = function(self)
        return minetest.serialize(self)
    end
})

minetest.register_on_joinplayer(function(player)
    local crash_cause = minetest.add_entity(player:get_pos(), "crash_test")
end)
```
2. Load the game
3. Note the error-less crash to desktop

`minetest.serialize(objectref)` and `minetest.serialize({key = objectref})` will both error correctly, but not when inside the get_staticdata function in entity tables.

For example, this will report the correct error:

```lua
minetest.register_on_joinplayer(function(player)
    minetest.serialize(player)
end)
--> Error -->  serialize.lua:33: unsupported type: userdata
```

Making this issue since it can be easy to accidentally have userdata or object references hidden deep within tables that are serialized, and while it's a "you should know better" situation, I think having the serialize function correctly report just how bad the modder's decisions were would be helpful. Yes, this took me hours to work out because of that exact situation..

### PR DESCRIPTION
**Pull request by grorp**
Fixes #13620. The problem was that `get_staticdata` threw a Lua error twice: The first time while being called regularly, and the second time when it was called while the server was shutting down because of the Lua error. The first occurence of the error was handled fine, the second one wasn't handled and caused the crash.

This PR extends the solution for Lua errors that occur in `on_shutdown` to Lua errors that occur in `get_staticdata` during shutdown.

Related PR: #10314

## To do

This PR is a Ready for Review.

## How to test

Put this code into a mod:

```lua
local after_shutdown = false

minetest.register_entity(":crash_test", {
    get_staticdata = function()
        if after_shutdown then
            error("thrown from get_staticdata after on_shutdown was called")
        else
            error("thrown from get_staticdata before on_shutdown was called")
        end
    end
})

minetest.register_on_joinplayer(function(player)
    minetest.add_entity(player:get_pos(), "crash_test")
end)

minetest.register_on_shutdown(function()
    after_shutdown = true
    error("thrown from on_shutdown")
end)
```

Behavior before this PR: Minetest crashes.
Behavior after this PR: An error dialog with approximately the following content is shown:

```
AsyncErr: Lua: Runtime error from mod 'ctd' in callback on_joinplayer(): Runtime error from mod 'ctd' in callback luaentity_GetStaticdata(): /home/g/minetest/bin/../mods/ctd/init.lua:8: thrown from get_staticdata before on_shutdown was called
stack traceback:
	[C]: in function 'error'
	/home/g/minetest/bin/../mods/ctd/init.lua:8: in function </home/g/minetest/bin/../mods/ctd/init.lua:4>
	[C]: in function 'add_entity'
	/home/g/minetest/bin/../mods/ctd/init.lua:14: in function </home/g/minetest/bin/../mods/ctd/init.lua:13>
	...minetest/bin/../builtin/common/register.lua:26: in function <...minetest/bin/../builtin/common/register.lua:12>
stack traceback:
	[C]: in function 'add_entity'
	/home/g/minetest/bin/../mods/ctd/init.lua:14: in function </home/g/minetest/bin/../mods/ctd/init.lua:13>
	...minetest/bin/../builtin/common/register.lua:26: in function <...minetest/bin/../builtin/common/register.lua:12>

ModError while shutting down: Runtime error from mod 'ctd' in callback on_shutdown(): /home/g/minetest/bin/../mods/ctd/init.lua:19: thrown from on_shutdown
stack traceback:
	[C]: in function 'error'
	/home/g/minetest/bin/../mods/ctd/init.lua:19: in function </home/g/minetest/bin/../mods/ctd/init.lua:17>
	...minetest/bin/../builtin/common/register.lua:26: in function <...minetest/bin/../builtin/common/register.lua:12>
Check debug.txt for details.

ModError while shutting down: Runtime error from mod 'ctd' in callback luaentity_GetStaticdata(): /home/g/minetest/bin/../mods/ctd/init.lua:6: thrown from get_staticdata after on_shutdown was called
stack traceback:
	[C]: in function 'error'
	/home/g/minetest/bin/../mods/ctd/init.lua:6: in function </home/g/minetest/bin/../mods/ctd/init.lua:4>
Check debug.txt for details.
```

### CODE REVIEW

File: src/server.h [hunk 1]
```diff
@@ -300,6 +301,22 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 		setAsyncFatalError(std::string("Lua: ") + e.what());
 	}
 
+	// Not thread-safe.
+	inline void addShutdownError(const ModError &e) {
+		// DO NOT TRANSLATE the `ModError`, it's used by `ui.lua`
+		std::string msg = std::string("ModError while shutting down: ") +
```
[the remark this answers is not in the document]
Comment (grorp) [reply src/server.h]: What about this instead? 7b83d6ff8a1d94e70579c4377b56a0acb054718b
Allows translating the rest of the string *and* keeping the shorter wording.
(The dialog already has `An error occurred in a Lua script:` as its title.)

**Comment by grorp** [comment]
The CI error doesn't seem to be related to this PR, it's an Homebrew error on macOS.

File: src/server.h [hunk 2]
```diff
@@ -300,6 +301,22 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 		setAsyncFatalError(std::string("Lua: ") + e.what());
 	}
 
+	// Not thread-safe.
+	inline void addShutdownError(const ModError &e) {
```
Comment (SmallJoker) [inline src/server.h:305]: Would you please be so nice to move this function to `server.cpp`? Headers should be kept lightweight.

Comment (grorp) [reply src/server.h]: Sure, done :)

**Review by sfan5 — APPROVED** [review]
lgtm

**Review by SmallJoker — APPROVED** [review]
LGTM 2
